src/Window.cpp

```diff
@@ -1432,33 +1695,205 @@ bool Window::moveForward(int distance) {
+        m_speedSlider->setValue(m_sliderValue*2.0);
+        break;
+    }
+    }
+    stats->addDistance(StatsEnum::COUNT_FULLFORWARD, distance);
+
+    return true;
+}
+
+bool Window::moveHalfForward(int distance) {
+// Non-positive distances aren't allowed
+if (distance < 1) {
+    return false;
+}
+// Special case for a wall directly in front of the mouse, else
+// the wall won't be detected until after the mouse starts moving
+//if (wallFront(0)) {
+//    return false;
+//}
+// Compute the number of allowable moves
+int moves = 1;
+int dis = distance/2 ;
+if( dis == 0 ) dis = 1;
+while (moves < dis) {
+    if (wallFront(moves)) {
+        stopRun("Stop running because of a wall!");
+        return false;
+    }
+    moves = dis;
+}
+m_movement = Movement::MOVE_HALFFORWARD;
+m_doomedToCrash = (moves != dis);
+m_movesRemaining = distance;
+
+if (m_startingLocation.first == 1 && m_startingLocation.second == 1) {
+    stats->startRun();
+}
+
+if(real)
+{
+switch(distance) {
+case 1:
+    m_speedSlider->setValue(m_sliderValue);
+    break;
+case 2 :
+    m_speedSlider->setValue(m_sliderValue*1.25);
+    break;
+case 3:
+    m_speedSlider->setValue(m_sliderValue*1.5);
+    break;
+default :
+    m_speedSlider->setValue(m_sliderValue*2.0);
+    break;
+}
+}
+
+// increase the stats by the distance that will be travelled
+stats->addDistance(StatsEnum::COUNT_HALFFORWARD, distance);
+return true;
+}
+
+bool Window::moveEdgeForward(int distance) {
+
+    // Non-positive distances aren't allowed
+    if (distance < 1) {
+        return false;
+    }
+    // Special case for a wall directly in front of the mouse, else
+    // the wall won't be detected until after the mouse starts moving
+    //if (wallFront(0)) {
+    //    return false;
+    // }
+    // Compute the number of allowable moves
+    int moves = 1;
+    //while (moves < distance) {
+    //    if (wallFront(moves)) {
+    //        break;
+    //    }
+    moves = distance;
+    //}
+    m_movement = Movement::MOVE_EDGEFORWARD;
+    m_doomedToCrash = (moves != distance);
+    m_movesRemaining = moves;
+
+
+    //if (m_startingLocation.first == 1 && m_startingLocation.second == 1) {
+    //    stats->startRun();
+    //}
+    // increase the stats by the distance that will be travelled
+    if(real)
+    {
+    switch(distance) {
+    case 1:
+        m_speedSlider->setValue(m_sliderValue);
+        break;
+    case 2 :
+        m_speedSlider->setValue(m_sliderValue*1.25);
+        break;
+    case 3:
+        m_speedSlider->setValue(m_sliderValue*1.5);
+        break;
+    default :
+        m_speedSlider->setValue(m_sliderValue*2.0);
+        break;
+    }
+    }
+    stats->addDistance(StatsEnum::COUNT_EDGEFORWARD, distance);
     return true;
 }
 
+void Window::turnBack() {
+    m_movement = Movement::TURN_BACK;
+    m_doomedToCrash = false;
+    m_movesRemaining = 0;
+    stats->addTurn(StatsEnum::COUNT_TURNBACK180);
+    if(real)
+    m_speedSlider->setValue(m_sliderValue);
```

Comment:
There's too much happening in these methods (and below). The movement methods should only be responsible for performing the movements. Setting slider values should be someone or something else's responsibility.